### PR TITLE
Port depthimage_to_laserscan to ROS2.

### PR DESCRIPTION
**Pull request by clalancette**
We need this to do 2D cartographer on turtlebot.

connects to ros2/ros2#342

CI:
* TB2 Linux [Build Status](http://ci.ros2.org/job/ci_turtlebot-demo_linux/24/)
* TB2 Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_turtlebot-demo_linux-aarch64/40/)

### CODE REVIEW

File: CMakeLists.txt
Comment (mikaelarguedas) [inline CMakeLists.txt]: Isn't that done by ament_target_dependencies below?

File: CMakeLists.txt
Comment (mikaelarguedas) [inline CMakeLists.txt]: I think we always build our libraries as SHARED. Otherwise we use ament_cmake_ros to control the behavior of building static or shared globally (same remark for every add_library call)

File: include/depthimage_to_laserscan/DepthImageToLaserScan.h
Comment (mikaelarguedas) [inline include/depthimage_to_laserscan/DepthImageToLaserScan.h]: Copyright format should be updated to pass copyright linter test, same for every file in this package

File: include/depthimage_to_laserscan/DepthImageToLaserScan.h
Comment (mikaelarguedas) [inline include/depthimage_to_laserscan/DepthImageToLaserScan.h]: please make whitespace fixes in a different commit

File: include/depthimage_to_laserscan/DepthImageToLaserScanROS.h
Comment (mikaelarguedas) [inline include/depthimage_to_laserscan/DepthImageToLaserScanROS.h]: Why change the include here but not in the previous file?

File: include/depthimage_to_laserscan/DepthImageToLaserScanROS.h
Comment (mikaelarguedas) [inline include/depthimage_to_laserscan/DepthImageToLaserScanROS.h]: Nit: indent is not consistent

File: test/DepthImageToLaserScanTest.cpp
Comment (mikaelarguedas) [inline test/DepthImageToLaserScanTest.cpp]: why changing the fixtures of the test is necessary here ?

File: test/DepthImageToLaserScanTest.cpp
Comment (mikaelarguedas) [inline test/DepthImageToLaserScanTest.cpp:161]: Nit: whitespace around operators

File: CMakeLists.txt
Comment (mikaelarguedas) [inline CMakeLists.txt]: Should the headers be installed as well? Should the libraries be installed in lib?

File: include/depthimage_to_laserscan/DepthImageToLaserScan.h
Comment (mikaelarguedas) [inline include/depthimage_to_laserscan/DepthImageToLaserScan.h]: if linting please update header guards

File: src/DepthImageToLaserScanROS.cpp
Comment (mikaelarguedas) [inline src/DepthImageToLaserScanROS.cpp:43]: start using logging macros instead ?

File: src/DepthImageToLaserScanROS.cpp
Comment (mikaelarguedas) [inline src/DepthImageToLaserScanROS.cpp]: I know it's redundant in all my reviews but please build your custom QOS profiles on top of the `rmw_qos_profile_sensor_data` profile that is designed exactly for this purpose

**Review by mikaelarguedas — COMMENTED** [review]
A few remarks about building and installing libraries and a few nitpicks otherwise that looks pretty good. I didn't review thoroughly though that's why this is just a review comment.

Not sure what the approach is here, I thought we did not change style of forked packages. If you do so please update the style everywhere and add the linter tests. Otherwise let's not make style changes at all. In any case please make style changes in a separate commit to make the review easier

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt]: Yeah, good call.  That was a leftover, I'll remove it.

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt]: I don't have a strong preference either way.  Looking around ros2, it is not totally consistent, so where I copied this from must not have had it.  I'll make it SHARED.

Comment (clalancette) [reply CMakeLists.txt]: Oh, I see.  I can't make it shared right now, because when I added the image_geometry library, I did not make that one SHARED.  I'll go and change that one as well.

Comment (mikaelarguedas) [reply CMakeLists.txt]: https://docs.google.com/a/osrfoundation.org/document/d/1jd8D4iR9Np4noE43IO1L6xwxNLqS0rYBMS5R7FFo_VA/edit?usp=sharing and https://github.com/ros2/navigation/pull/1 give good examples of the steps for porting. We faced the same thing when trying to use image_geometry and some camera stuff with @Kukanani recently.

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt]: I wasn't quite sure about that one.  This is a "leaf node", so what would include/link against it?  That being said, I'm not against doing this, I just wasn't sure it was necessary.

File: include/depthimage_to_laserscan/DepthImageToLaserScan.h
Comment (clalancette) [reply include/depthimage_to_laserscan/DepthImageToLaserScan.h]: I don't think we should necessarily lint this, so that we can pull updates from upstream later (if necessary).  The only formatting changes I did were whitespace, which I'll probably remove.

File: include/depthimage_to_laserscan/DepthImageToLaserScanROS.h
Comment (clalancette) [reply include/depthimage_to_laserscan/DepthImageToLaserScanROS.h]: That was just a mistake.  I'll fix it.

File: include/depthimage_to_laserscan/DepthImageToLaserScanROS.h
Comment (clalancette) [reply include/depthimage_to_laserscan/DepthImageToLaserScanROS.h]: Good catch, will fix.

File: test/DepthImageToLaserScanTest.cpp
Comment (clalancette) [reply test/DepthImageToLaserScanTest.cpp]: The problem is that header.stamp doesn't have a fromNSec() method anymore.  I just filled out the structure with 0 to make my life easy, since it isn't used in the rest of the test.

File: src/DepthImageToLaserScanROS.cpp
Comment (clalancette) [reply src/DepthImageToLaserScanROS.cpp]: Sure, will do.

File: src/DepthImageToLaserScanROS.cpp
Comment (clalancette) [reply src/DepthImageToLaserScanROS.cpp:43]: Yeah, I guess I'll switch to that now.

Comment (mikaelarguedas) [reply src/DepthImageToLaserScanROS.cpp:43]: actually maybe I going to quick on this, I think it's fine to keep it as is until we have logging macros fully implemented in rclcpp

Comment (clalancette) [reply src/DepthImageToLaserScanROS.cpp:43]: That's fine with me.  I was having trouble making it work right now anyway.  Once we get that more fully implemented, and we are past beta2, I'll make a sweep back through the code and fix a lot of these.

**Comment by clalancette** [comment]
I ended up reverting the whitespace changes to a) keep the diff down, and b) keep us closer to upstream.  I'll open a PR with style/whitespace changes on the upstream repository.

File: CMakeLists.txt
Comment (mikaelarguedas) [reply CMakeLists.txt]: Anything wanting to load it in a container or an executor to leverage zero-copy for example. That being all the ported code so far is ported in a ros1 way that doesn't allow to use it in that context just yet

**Comment by mikaelarguedas** [comment]
> I ended up reverting the whitespace changes to a) keep the diff down, and b) keep us closer to upstream. I'll open a PR with style/whitespace changes on the upstream repository.

Sounds good, there is still a bunch of semi-colon removal, indents and whitespaces in this PR though

**Comment by clalancette** [comment]
We have to do the semi-colon removal; it won't compile with -Wpedantic otherwise.  The rest of the whitespace changes are just adding a \n at the end of the file; I think we can leave those, as they won't really make it difficult to merge later.  I'll revert the indentation changes as well.  I've opened up ros-perception/depthimage_to_laserscan#24 for the whitespace changes.

**Comment by mikaelarguedas** [comment]
It will compile but with a warning. I just think that all of these are improving upstream and not ROS2 specific so they should land in the upstream PR. Otherwise they should at least be in a separate commit

**Comment by clalancette** [comment]
Yeah, I'll split them out into a separate commit; that will go away once the upstream PR lands.

**Comment by clalancette** [comment]
New CI to test out the shared stuff:
* TB2 Linux [Build Status](http://ci.ros2.org/job/ci_turtlebot-demo_linux/34/)
* TB2 Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_turtlebot-demo_linux-aarch64/48/)

**Comment by clalancette** [comment]
@ros2/team I think this one is ready to go now; anyone want to do a final review/approval so I can merge?

File: src/DepthImageToLaserScanROS.cpp
Comment (mikaelarguedas) [inline src/DepthImageToLaserScanROS.cpp]: Same question as https://github.com/ros2/ros_astra_camera/pull/17#discussion_r121443254 regarding performance improvement compared to the default sensor_data qos profile

File: src/DepthImageToLaserScanROS.cpp
Comment (mikaelarguedas) [inline src/DepthImageToLaserScanROS.cpp:65]: Using get_parameter_or() make each of these a one-liner

File: test/DepthImageToLaserScanTest.cpp
Comment (mikaelarguedas) [inline test/DepthImageToLaserScanTest.cpp]: Nit: c++ header imported in between c headers?

File: src/DepthImageToLaserScanROS.cpp
Comment (clalancette) [reply src/DepthImageToLaserScanROS.cpp:65]: Oh yeah, definitely.  Now that I know about it, good call, will fix.

File: test/DepthImageToLaserScanTest.cpp
Comment (clalancette) [reply test/DepthImageToLaserScanTest.cpp]: Ah, right.  To be honest, I still don't quite understand the order of C vs C++ includes that we prefer.  I didn't see it in the style guide.  Should I move this to the top, then?

File: src/DepthImageToLaserScanROS.cpp
Comment (clalancette) [reply src/DepthImageToLaserScanROS.cpp:65]: Actually, it doesn't end up being a one-liner since I still need to declare the variables.  Further, get_parameter_or() doesn't seem to have a templated version for float, so I need to change the types of all of the variables to double, and then I'd have to cast back to float for the call to dtl_->set*().  Overall, it doesn't seem to me that changing this to get_parameter_or() is a win.  Thoughts?

File: test/DepthImageToLaserScanTest.cpp
Comment (mikaelarguedas) [reply test/DepthImageToLaserScanTest.cpp]: we follow the [google style guide](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes), our developer guide only highlight where we differ from it.

-    header of the current file
-    C system files.
-    C++ system files.
-    Other libraries' .h files.
-    Your project's .h files.

File: src/DepthImageToLaserScanROS.cpp
Comment (mikaelarguedas) [reply src/DepthImageToLaserScanROS.cpp:65]: > doesn't seem to have a templated version for float

We don't support float parameters at all AFAIK (https://github.com/ros2/rcl_interfaces/blob/819485ed7ba2782652ecf93ab1c1a1c6703cdeb7/rcl_interfaces/msg/ParameterType.msg)
. So there will be a cast (implicit of explicit) in any case.

We can keep the `get_parameter()` call as is for now but we should be aware of that cast happening.

File: test/DepthImageToLaserScanTest.cpp
Comment (clalancette) [reply test/DepthImageToLaserScanTest.cpp]: Thanks.  I moved my addition down (underneath the C system ones).  I did not rearrange the rest of the headers, in order to keep differences with upstream down.

File: src/DepthImageToLaserScanROS.cpp
Comment (clalancette) [reply src/DepthImageToLaserScanROS.cpp]: I just tried it out a few times with the default sensor_data QoS profile, and it seemed to work.  So I'll switch to using that instead.

File: src/DepthImageToLaserScanROS.cpp
Comment (mikaelarguedas) [inline src/DepthImageToLaserScanROS.cpp:47]: do we need `:node_(node)`? Isn't the node provided as argument enough ?

File: CMakeLists.txt
Comment (mikaelarguedas) [inline CMakeLists.txt]: Nit: please use several lines if passing several libraries to link against

**Review by mikaelarguedas — APPROVED** [review]
lgtm except one comment that is not preventing from merging

File: src/DepthImageToLaserScanROS.cpp
Comment (clalancette) [reply src/DepthImageToLaserScanROS.cpp:47]: This is a pattern I've seen in some other ROS nodes, and upon reflection, I thought it was a good idea.  The reason is to keep a reference to the node in-scope.  Imagine you have the following (pseudo-)code:

```
void init()
{
  rclcpp::node::Node::SharedPtr node = rclcpp::node::Node::make_shared("depthimage_to_laserscan");

  depthimage_to_laserscan::DepthImageToLaserScanROS dtl(node);
}

int main()
{
    init();
}
```

In this case, when init() completes, node goes out of scope and is destroyed.  While this doesn't quite matter right now (the DepthImageToLaserScanROS class never references node outside of the constructor), it seems to me that it is a good idea to hold onto a reference since we may need it in the future.

File: CMakeLists.txt
Comment (clalancette) [reply CMakeLists.txt]: Will fix.

**Comment by clalancette** [comment]
All right, I fixed the nit, and rebased the branch.  I'm now going to merge.  Thanks!

**Pull request merged**
